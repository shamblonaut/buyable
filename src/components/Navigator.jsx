import { Link } from "react-router-dom";
import styled from "styled-components";
import { Home, ShoppingBag, ShoppingCart } from "lucide-react";

const Bar = styled.nav`
  position: fixed;
  bottom: 0;
  width: 100%;

  display: flex;
  justify-content: space-around;
  padding: 16px;
  border-top: 2px solid black;
  background-color: white;

  & .lucide {
    color: black;
  }
`;

const Navigator = () => (
  <Bar>
    <Link to="/">
      <Home />
    </Link>
    <Link to="/shop">
      <ShoppingBag />
    </Link>
    <Link to="/cart">
      <ShoppingCart />
    </Link>
  </Bar>
);

export default Navigator;
