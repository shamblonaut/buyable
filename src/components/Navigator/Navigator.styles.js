import styled from "styled-components";

export const Bar = styled.nav`
  position: fixed;
  bottom: 0;
  width: 100%;
  height: 64px;

  display: flex;
  justify-content: center;

  background-color: var(--color-accent-dark);
  border-top: 3px solid var(--color-dark);
`;

export const Tabs = styled.div`
  height: 100%;
  width: 100%;
  max-width: 1280px;
  display: flex;
  justify-content: space-around;
  align-items: center;

  & .lucide {
    color: var(--color-light);
  }
`;
