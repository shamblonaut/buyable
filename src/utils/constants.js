export const AppPosition = Object.freeze({
  HOME: 0,
  SHOP: 1,
  CART: 2,
});
