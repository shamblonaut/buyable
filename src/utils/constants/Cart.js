export const CartAction = Object.freeze({
  ADDED: "added_to_cart",
  REMOVED: "removed_from_cart",
  INCREMENTED: "incremented_cart_quantity",
  DECREMENTED: "decremented_cart_quantity",
  EDITED: "edited_cart_quantity",
  CLEARED: "cleared_cart",
});
