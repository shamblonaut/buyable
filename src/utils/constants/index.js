import { AppPosition } from "./App";

export { AppPosition };
