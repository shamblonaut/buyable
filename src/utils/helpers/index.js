import { getPositionRoute } from "./App";
import { getRatingArray } from "./Rating";
import { toTitleCase } from "./Misc";

export { getPositionRoute, getRatingArray, toTitleCase };
