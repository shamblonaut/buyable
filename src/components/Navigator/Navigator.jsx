import { useContext } from "react";
import { Home, ShoppingBag, ShoppingCart } from "lucide-react";
import PropTypes from "prop-types";

import { CartContext } from "@/contexts";
import { AppPosition } from "@/utils/constants";

import { Bar, Tabs, PageLink, CartIndicator } from "./Navigator.styles";

const Navigator = ({ appPosition }) => {
  const { cart } = useContext(CartContext);

  const cartCount = Object.values(cart).reduce(
    (sum, current) => sum + current,
    0,
  );

  return (
    <Bar>
      <Tabs>
        <PageLink
          to="/"
          aria-label="Go to Home page"
          $active={appPosition === AppPosition.HOME}
        >
          <Home />
        </PageLink>
        <PageLink
          to="/shop"
          aria-label="Go to Shop page"
          $active={appPosition === AppPosition.SHOP}
        >
          <ShoppingBag />
        </PageLink>
        <PageLink
          to="/cart"
          aria-label="Go to Cart page"
          $active={appPosition === AppPosition.CART}
        >
          <ShoppingCart />
          {cartCount > 0 && (
            <CartIndicator>{cartCount < 100 ? cartCount : "99+"}</CartIndicator>
          )}
        </PageLink>
      </Tabs>
    </Bar>
  );
};

Navigator.propTypes = {
  appPosition: PropTypes.oneOf(Object.values(AppPosition)),
};

export default Navigator;
