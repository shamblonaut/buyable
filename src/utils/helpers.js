import { AppPosition } from "./constants";

export const getRatingArray = (rating) => {
  if (rating > 5.0) {
    throw new Error("Invalid rating");
  }

  let ratingArray = [];
  for (let i = 0; i < 5; i++) {
    const remainder = rating - i;
    ratingArray.push(
      remainder > 1 ? 1 : remainder > 0 ? Number(remainder.toFixed(1)) : 0,
    );
  }

  return ratingArray;
};

export const getPositionRoute = (appPosition) => {
  switch (appPosition) {
    case AppPosition.HOME:
      return "/";
    case AppPosition.SHOP:
      return "/shop";
    case AppPosition.CART:
      return "/cart";
  }
};

export const toTitleCase = (string) => {
  let result = "";
  for (const word of string.split(" ")) {
    result += word[0].toUpperCase() + word.substring(1) + " ";
  }
  return result.trim();
};
