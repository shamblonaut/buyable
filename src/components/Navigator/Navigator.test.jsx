import { describe, it, expect } from "vitest";
import { render, screen, within } from "@testing-library/react";

import { BrowserRouter } from "react-router-dom";

import { AppPosition } from "@/utils/constants";
import Navigator from "./Navigator";

const mockCartCount = 16;

describe("Navigator component", () => {
  it("renders correctly", () => {
    const { container } = render(
      <BrowserRouter>
        <Navigator appPosition={AppPosition.HOME} cartCount={mockCartCount} />,
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it("highlights currently active page", () => {
    render(
      <BrowserRouter>
        <Navigator appPosition={AppPosition.SHOP} cartCount={mockCartCount} />
      </BrowserRouter>,
    );

    expect(
      screen.getByRole("link", { name: "Go to Home page" }).firstChild,
    ).not.toHaveStyle({
      color: "var(--color-light)",
    });
    expect(
      screen.getByRole("link", { name: "Go to Shop page" }).firstChild,
    ).toHaveStyle({
      color: "var(--color-light)",
    });
    expect(
      screen.getByRole("link", { name: "Go to Cart page" }).firstChild,
    ).not.toHaveStyle({
      color: "var(--color-light)",
    });
  });

  it("shows current no. of items in cart within cart page link", () => {
    render(
      <BrowserRouter>
        <Navigator appPosition={AppPosition.SHOP} cartCount={mockCartCount} />
      </BrowserRouter>,
    );

    expect(
      within(screen.getByRole("link", { name: "Go to Cart page" })).queryByText(
        mockCartCount.toString(),
      ),
    ).toBeInTheDocument();
  });

  it("links to respective pages correctly", () => {
    render(
      <BrowserRouter>
        <Navigator appPosition={AppPosition.HOME} cartCount={mockCartCount} />
      </BrowserRouter>,
    );

    expect(
      screen.getByRole("link", { name: "Go to Home page" }),
    ).toHaveAttribute("href", "/");
    expect(
      screen.getByRole("link", { name: "Go to Shop page" }),
    ).toHaveAttribute("href", "/shop");
    expect(
      screen.getByRole("link", { name: "Go to Cart page" }),
    ).toHaveAttribute("href", "/cart");
  });
});
