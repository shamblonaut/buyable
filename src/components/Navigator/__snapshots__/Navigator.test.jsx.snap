// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navigator component > renders correctly 1`] = `
<div>
  <nav
    class="sc-bRKDuR iMsCxZ"
  >
    <div
      class="sc-hvigdm hVLfAZ"
    >
      <a
        aria-label="Go to Home page"
        class="sc-fhHczv kRcGpD"
        data-discover="true"
        href="/"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-house"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 21v-8a1 1 0 0 0-1-1h-4a1 1 0 0 0-1 1v8"
          />
          <path
            d="M3 10a2 2 0 0 1 .709-1.528l7-5.999a2 2 0 0 1 2.582 0l7 5.999A2 2 0 0 1 21 10v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
          />
        </svg>
      </a>
      <a
        aria-label="Go to Shop page"
        class="sc-fhHczv jNLQtO"
        data-discover="true"
        href="/shop"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-shopping-bag"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 10a4 4 0 0 1-8 0"
          />
          <path
            d="M3.103 6.034h17.794"
          />
          <path
            d="M3.4 5.467a2 2 0 0 0-.4 1.2V20a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V6.667a2 2 0 0 0-.4-1.2l-2-2.667A2 2 0 0 0 17 2H7a2 2 0 0 0-1.6.8z"
          />
        </svg>
      </a>
      <a
        aria-label="Go to Cart page"
        class="sc-fhHczv jNLQtO"
        data-discover="true"
        href="/cart"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-shopping-cart"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="8"
            cy="21"
            r="1"
          />
          <circle
            cx="19"
            cy="21"
            r="1"
          />
          <path
            d="M2.05 2.05h2l2.66 12.42a2 2 0 0 0 2 1.58h9.78a2 2 0 0 0 1.95-1.57l1.65-7.43H5.12"
          />
        </svg>
        <div
          class="sc-ggWZvA iRPAEk"
        >
          55
        </div>
      </a>
    </div>
  </nav>
</div>
`;
